File: practice_datastructure/112. Path Sum.cpp
class Solution {
public:
    bool hasPathSum(TreeNode* root, int target) {
        if(root == NULL) return false;
        target -= root->val;
        queue<pair<TreeNode*, int>>q;
        q.push({root, target});

        while(!q.empty()){
            pair<TreeNode*, int>top = q.front(); q.pop();
            TreeNode* tmp = top.first;
            int remain = top.second;
            if(tmp->left == NULL && tmp->right == NULL && remain == 0) return true;
            if(tmp->left!=NULL) q.push({tmp->left, remain-tmp->left->val});
            if(tmp->right!=NULL) q.push({tmp->right, remain-tmp->right->val});
        }
        return false;
    }
};

File: PastYears/practice_datastructure/112. Path Sum.java
class Solution {

    public boolean dfs(TreeNode root, int remain){
        if(root == null) return false;
        if(root != null && root.left == null && root.right == null){
            if(remain == root.val) return true;
            return false;
        } 
        boolean left = dfs(root.left, remain-root.val);
        if(left) return true;
        boolean right = dfs(root.right, remain-root.val);
        return left||right;
    }

    public boolean hasPathSum(TreeNode root, int target) {
        if(root == null) return false;
        return dfs(root, target);
    }
}


class Solution {
    public boolean hasPathSum(TreeNode root, int target) {
        if(root == null) return false;
        if(root.left == null && root.right == null){
            if(target == root.val) return true;
            return false;
        } 
        boolean left = hasPathSum(root.left, target-root.val);
        if(left) return true;
        boolean right = hasPathSum(root.right, target-root.val);
        return left||right;
    }
}

class Solution {
    public boolean dfs(TreeNode root, int target, int count){
        if(root == null) return false;
        count += root.val;
        if(root.left == null && root.right == null) return count == target;

        boolean left = dfs(root.left, target, count);
        if(left) return true;
        boolean right = dfs(root.right, target, count);
        return left||right;
    }

    public boolean hasPathSum(TreeNode root, int target) {
        if(root == null) return false;
        return dfs(root, target, 0);
    }
}
